1-Draft/RFCnnnn-ForEach-Parallel-Implementation.md

```diff
@@ -0,0 +1,110 @@
+        Start-Sleep -Seconds 1
+        Write-Output "Output $item"
+    }
+}
+
+```
+
+This will run the script block with each value in the `$list` array, in parallel using workflow jobs.
+However, workflow is not supported in PowerShell Core 6, partly because it is a Windows only solution but also because it is cumbersome to use.
+In addition the workflow implementation is very heavy weight, using lots of system resources.  
+
+This is a proposal to re-implement `foreach -parallel` in PowerShell Core, using PowerShell's support for concurrency via Runspaces.
+It is similar to the [ThreadJob module](https://www.powershellgallery.com/packages/ThreadJob/1.1.2) except that it becomes part of the PowerShell language via `foreach -parallel`.
+
+## Motivation
+
+    As a PowerShell User,
+    I can do simple fan-out concurrency from within the language, without having to obtain and load a separate module or deal with PowerShell jobs.
+
+## Specification
+
+The PowerShell `foreach -parallel` language keyword will be re-implemented to perform invoke script blocks in parallel, similar to how it works for workflow functions except that script blocks will be invoked on threads within the same process rather than in workflow jobs running in separate processes.
+The default behavior is to fan-out script block execution to multiple threads, and then wait for all threads to finish.
+However, a `-asjob` switch will also be supported that returns a PowerShell job object for asynchronous use.
+If the number of foreach iterations exceed the throttle limit value, then only the throttle limit number of threads are created at a time and the rest are queued until a running thread becomes available.
+
+### Supported foreach parameters
+
+- `-parallel`
+- `-throttlelimit`
+- `-timeout`
+- `-asjob`
+
+### P0 Features
+
+- `foreach -parallel` fans out script block execution to threads, along with a bound single foreach iteration value
+
+- `-throttlelimit` parameter value specifies the maximum number of threads that can run at one time
+
+- `-timeout` parameter value specifies a maximum time to wait for all iterations to complete, after which 'stop' will be called on all running script blocks to terminate execution
+
+- `-asjob` switch causes foreach to return a PowerShell job object that is used to asynchronously monitor execution
+
+- When a job object is returned, it will be compatible with all relevant job cmdlets
+
+- All script blocks running in parallel will run isolated from each other.
+Only foreach iteration objects will be passed to the parallel script block.
+
+### Data stream handling
+
+`foreach -parallel` will use normal PowerShell pipes to return various data streams.
+Data will be returned in order received.  
+Except when `-asjob` switch is used, in which case a single job object is returned.
+The returned job object will contain an array of child jobs that represent each iteration of the foreach.

```

Comment:
> The returned job object will contain an array of child jobs that represent each iteration of the foreach.

So when used together with with the `-throttlelimit` this will return an array of all child jobs, and the number of child jobs exceed the throttle limit value, will those have status 'Waiting'? *In relation to the section 'Specification' too.*

Reply:
Yes, I anticipate creating a single child job for each loop iteration when using the `-asjob` switch.  The state of the job will be `NotStarted` when queued due to throttle limit.  This could become a resource problem if a foreach loop is created with 10K(!) iterations.  Implementation should use lazy initialization and disposal at the earliest opportunity strategies to minimize resource usage.